packages/netsuite-adapter/src/filters/fix_order_problems_in_workbooks.ts

```diff
@@ -0,0 +1,199 @@
+const { awu } = collections.asynciterable
+
+const addWorkbookToolsToRecord = (
+  instance: InstanceElement,
+  workbookToolsRecord: Record<string, Values>,
+  name: string
+): void => {
+  if (Array.isArray(instance.value[name])) {
+    instance.value[name].forEach((tool:unknown) => {
+      if (values.isPlainObject(tool)) {
+        Object.values(tool)
+          .filter(values.isPlainObject)
+          .forEach((obj: Values) => {
+            if (_.isString(obj[SOAP_SCRIPT_ID])) {
+              workbookToolsRecord[obj[SOAP_SCRIPT_ID]] = obj
+            }
+          })
+      }
+    })
+  }
+}
+
+const createRecordOfPivotsChartsAndDsLinks = async (
+  elementsSource: ReadOnlyElementsSource,
+  elemID: ElemID,
+): Promise<Record<string, Values>> => {
+  const oldWorkbookToolsRecord: Record<string, Values> = {}
+  const oldInstance = await elementsSource.get(elemID)
+  if (isInstanceElement(oldInstance)) {
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, PIVOTS)
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, CHARTS)
+    addWorkbookToolsToRecord(oldInstance, oldWorkbookToolsRecord, DATASET_LINKS)
+  }
+  return oldWorkbookToolsRecord
+}
+
+const isSameXmlValues = (xml1: string, xml2: string): boolean => {
+  const values1 = parse(xml1, {
+    attributeNamePrefix: ATTRIBUTE_PREFIX,
+    ignoreAttributes: false,
+    tagValueProcessor: val => decode(val),
+  })
+  const values2 = parse(xml2, {
+    attributeNamePrefix: ATTRIBUTE_PREFIX,
+    ignoreAttributes: false,
+    tagValueProcessor: val => decode(val),
+  })
+  return _.isEqual(values1, values2)
+}
+
+const getInnerXmlTitle = (workbookTool: Values): string | undefined =>
+  INNER_XML_TITLES.find(title => title in workbookTool)
+
+const compareAndAssignInnerXmls = (newTool: Values, oldTool: Values): void => {
+  const innerXmlTitle = getInnerXmlTitle(newTool)
+  if (innerXmlTitle === undefined) {
+    return
+  }
+  const innerXmlNew = newTool[innerXmlTitle]
+  const innerXmlOld = oldTool[innerXmlTitle]
+  if (!_.isString(innerXmlNew) || !_.isString(innerXmlOld)) {
+    return
+  }
+  if (isSameXmlValues(innerXmlNew, innerXmlOld)) {
+    newTool[innerXmlTitle] = oldTool[innerXmlTitle]
+  }
+}
+
+const adjustWorkbookToolsOrder = (workbookTools: Value[], oldWorkbookToolsRecord: Record<string, Values>): void => {
+  workbookTools
+    .filter(values.isPlainObject)
+    .flatMap(obj => Object.values(obj))
+    .forEach((workbookTool: Values) => {
+      if (_.isArray(workbookTool[DATASETS])) {
+        workbookTool[DATASETS].sort()
+      }
+      if (workbookTool[SOAP_SCRIPT_ID] in oldWorkbookToolsRecord) {
+        const oldWorkbookTool = oldWorkbookToolsRecord[workbookTool[SOAP_SCRIPT_ID]]
+        compareAndAssignInnerXmls(workbookTool, oldWorkbookTool)
+      }
+    })
+}
+
+const discardUnrelevantChanges = (
+  newInstanceValues: Values,
+  oldTools: Record<string, Values>,
+): void => {
+  INNER_ARRAY_NAMES.forEach((arrName: string) => {
```

Comment:
nit - type announcing is redundant when then type of `INNER_ARRAY_NAMES` is well defined:
```suggestion
  INNER_ARRAY_NAMES.forEach(arrName => {
```